### PR TITLE
Disallow selection of ownerless nodes

### LINKED ISSUE #92187: Node3D editor allows selection of ownerless nodes.
### Tested versions

40b4130c93d08235a60996d29e5869a22b6ae53d

### System information

Godot 4.3

### Issue description

The Node3D editor allow the user to click on nodes with selectable gizmos such as MeshInstance3D even if they are marked as not having an owner. As a result, they will be selected but with no corresponding selection to be found in the scene tree editor. This behaviour feels very incorrect and can be a major discontinuity when creating certain types of editor tool scripts. I have attached a project which demonstrates a tool script which uses a PackedScene property to create an ownerless node which acts as a preview.

### Steps to reproduce

Open the project and click on the character mesh. It should be selected, but nothing corresponding will appear in the scene tree editor.

### Minimal reproduction project (MRP)

[instance_scenes.zip](https://github.com/godotengine/godot/files/15384711/instance_scenes.zip)

### PR DESCRIPTION
**Pull request by SaracenOne**
Closes #92187

This changes the behaviour of node selection when selecting ownerless nodes in the 3D node editor. Now, instead of simply selecting the ownerless node, it will walk its parents and attempt to find the first node which is valid for selection.

### CODE REVIEW

**Comment by ryevdokimov** [comment]
I agree that having nothing shown as selected in the SceneTree is confusing, but maybe alternative it to show some kind transient node selection that is denoted somehow (like yellow being nodes owned by another scene in the case of inherited scenes). My concern now is that you can't access and modify the ownerless nodes properties in the inspector, which might be desirable in some circumstance.

**Comment by SaracenOne** [comment]
It's possible there is use for it, but looking everything which is generally considered standard operation in Godot, I really can't think of what that might be, and as pointed out in the issue, the current behaviour directly conflicts with a otherwise perfectly viable data-driven workflow. If the goal was to provide something similar with the behaviour of instanced scenes with the editable scene flag, I think at the very least there would be a requirement to flag an ownerless node as editable too before it can be selected.

**Comment by KoBeWi** [comment]
This is consistent with 2D editor. The equivalent code is here:
https://github.com/godotengine/godot/blob/8e2141eac534f6984bb0bdbcefbd17de27ae0993/editor/plugins/canvas_item_editor_plugin.cpp#L710
Maybe `get_deepest_editable_node()` could be useful for your code too.

File: editor/plugins/node_3d_editor_plugin.cpp [hunk 1]
```diff
@@ -1529,23 +1550,35 @@ bool Node3DEditorViewport ::_is_node_locked(const Node *p_node) {
...
+	// If a node has an invalid owner, recursively check their parents until a valid node is found.
+	for (int i = 0; i < potential_selection_results.size(); i++) {
+		Node3D *node = potential_selection_results[i].item;
+		while (1) {
+			if (node == nullptr || node == edited_scene->get_parent()) {
+				break;
+			} else {
+				Node *node_owner = node->get_owner();
+				if (node == edited_scene || node_owner == edited_scene || (node_owner != nullptr && edited_scene->is_editable_instance(node_owner))) {
+					if (selection_results.find(node) == -1) {
```
Comment (KoBeWi) [inline editor/plugins/node_3d_editor_plugin.cpp:1568]:
```suggestion
					if (!selection_results.has(node)) {
```

**Review by KoBeWi — APPROVED** [review]

**Comment by caimantilla** [comment]
Aaah, I thought I was going crazy the other day! This happens with my spawner class, which instantiates the character scene to different positions for preview (orientation, entrance origin, destination, etc.)
Very keen on seeing this merged.

**Comment by akien-mga** [comment]
Thanks!

**Comment by mihe** [comment]
I stumbled upon a regression that seems to have been caused by this PR, where you're no longer able to use the list of selectable (overlapping) nodes feature in the 3D editor.

I put up a fix in #94387, in case anyone wants to take a look.

**Comment by mihe** [comment]
I'm also noticing that the code for this list selection feature now differs between 3D and 2D, where the 3D list selection now incorporates the logic from `CanvasItemEditor::_find_canvas_items_in_rect` that was [mentioned above](https://github.com/godotengine/godot/pull/92188#issuecomment-2125010129) but the 2D list selection still just overwrites `selection_results` with the results from `CanvasItemEditor::_get_canvas_items_at_pos`, much like what the 3D editor did before.

Is that something that should be addressed?

**Pull request merged**